.github/actions/trun/run.ts

```diff
@@ -0,0 +1,132 @@
+import * as esbuild from "./deps/esbuild.ts"
+import { denoPlugin } from "./deps/esbuild_deno_loader.ts"
+import PQueue from "./deps/pqueue.ts"
+import { Browser } from "./deps/puppeteer.ts"
+import { deferred } from "./deps/std/async.ts"
+import { Buffer, readLines } from "./deps/std/io.ts"
+import * as path from "./deps/std/path.ts"
+import { readerFromStreamReader, writeAll } from "./deps/std/streams.ts"
+
+export interface RunOptions {
+  paths: (readonly [dir: string, fileName: string])[]
+  concurrency: number
+  runner: (dir: string, fileName: string) => Promise<void>
+}
+
+export async function run({ paths, concurrency, runner }: RunOptions) {
+  const runQueue = new PQueue({ concurrency })
+  runQueue.addAll(paths.map(([dir, fileName]) => () => runner(dir, fileName)))
+  await runQueue.onIdle()
+}
+
+export interface RunWithBrowserOptions {
+  createBrowser: () => Promise<Browser>
+  importMapURL?: URL
+  results: [fileName: string, exitCode: number][]
+}
+
+export async function runWithBrowser(
+  { createBrowser, importMapURL, results }: RunWithBrowserOptions,
+) {
+  const browser = await createBrowser()
+  const currentDir = new URL(import.meta.url).pathname.split("/").slice(0, -1).join("/")
+  const consoleJS = new TextDecoder().decode(await Deno.readFile(`${currentDir}/console.js`))
+
+  return (async (dir: string, fileName: string) => {
+    console.log(`running ${fileName}`)
+    const outputQueue = new PQueue({ concurrency: 1, autoStart: false })
+
+    outputQueue.add(() => console.log(`${fileName} output:`))
+
+    const page = await browser.newPage()
+    const result = await esbuild.build({
+      plugins: [
+        denoPlugin({
+          importMapURL,
+        }) as any,
+      ],
+      entryPoints: [path.join(dir, fileName)],
+      bundle: true,
+      write: false,
+      format: "esm",
+    })
+
+    const code = wrapCode(`${consoleJS}\n${result.outputFiles[0]?.text!}`)
+
+    await page.exposeFunction("__trun_injected_log", (...args: unknown[]) => {
+      outputQueue.add(() => console.log(args))
```

Comment:
Instead of pushing to a `PQueue`, can you write to a `Buffer`?

Reply:
Like `JSON.stringify` each line into a buffer and then when reading `JSON.parse`? Reason its done this way is because you lose the "console richness" if you just convert it into a string and throw it into a buffer and then read from it.

Reply:
The arguments you receive from `console.js` are of the form `(log: string, level: number)`; the first argument has color baked in already, and the second argument indicates stdout/stderr (which isn't really relevant here). So if you `textEncoder.encode(log)`, and write that into a `Buffer`, when you print that buffer out in the end it'll have all the relevant color and formatting.